File: Source/Events/Commands/AwardingCmds.cs
using System.Linq;
using System.Reflection;
using System.Text;
using WCSharp.Api;
using static WCSharp.Api.Common;

public static class AwardingCmds
{
    private static readonly string[] CategoryColors =
    {
        Colors.COLOR_TURQUOISE, Colors.COLOR_GOLD,    Colors.COLOR_PURPLE,
        Colors.COLOR_LAVENDER,  Colors.COLOR_YELLOW,  Colors.COLOR_PINK,
        Colors.COLOR_RED,       Colors.COLOR_GREEN,   Colors.COLOR_CYAN,
    };

    private static readonly StringBuilder _sb = new StringBuilder();
    private static readonly StringBuilder _inner = new StringBuilder();

    /// <summary>
    /// Awards the resolved player with the given reward. Use ?award help to see valid awards grouped by category.
    /// </summary>
    public static void Awarding(player player, string[] args)
    {
        if (args[0] == "") return;

        var award = args[0].ToLower();

        if (award == "help")
        {
            AwardingHelp(player, args.Length > 1 ? args[1] : "");
            return;
        }

        if (award == "all")
        {
            CommandsManager.ResolvePlayerId(args.Length > 1 ? args[1] : "", kitty => AwardAll(kitty.Player));
            return;
        }

        var foundAward = FindAwardName(award);
        if (foundAward == null)
        {
            player.DisplayTimedTextTo(3.0f, $"{Colors.COLOR_YELLOW_ORANGE}No valid award found for: |r{Colors.HighlightString(award)} {Colors.COLOR_YELLOW_ORANGE}— try ?award help|r");
            return;
        }

        CommandsManager.ResolvePlayerId(args.Length > 1 ? args[1] : "", kitty =>
        {
            if (kitty == null) return;
            AwardManager.GiveReward(kitty.Player, foundAward);
            player.DisplayTimedTextTo(3.0f, $"{Colors.COLOR_YELLOW_ORANGE}Awarded {Colors.HighlightString(foundAward)} to {Colors.PlayerNameColored(kitty.Player)}|r");
        });
    }

    /// <summary>
    /// Removes the specified award from the resolved player. Args: [award name] [player?]
    /// </summary>
    public static void RemovingAward(player player, string[] args)
    {
        if (args[0] == "")
        {
            player.DisplayTimedTextTo(3.0f, $"{Colors.COLOR_YELLOW_ORANGE}Usage: removeaward [award name] [player?]|r");
            return;
        }

        var foundAward = FindAwardName(args[0].ToLower());
        if (foundAward == null)
        {
            player.DisplayTimedTextTo(3.0f, $"{Colors.COLOR_YELLOW_ORANGE}No valid award found for: |r{Colors.HighlightString(args[0])} {Colors.COLOR_YELLOW_ORANGE}— try ?award help|r");
            return;
        }

        CommandsManager.ResolvePlayerId(args.Length > 1 ? args[1] : "", kitty =>
        {
            if (kitty == null) return;
            AwardManager.RemoveReward(kitty.Player, foundAward);
            player.DisplayTimedTextTo(3.0f, $"{Colors.COLOR_YELLOW_ORANGE}Removed {Colors.HighlightString(foundAward)} from {Colors.PlayerNameColored(kitty.Player)}|r");
        });
    }

    /// <summary>
    /// Sets the specified game stat for the resolved player. Args: [stat] [value] [player?]. Use ?stat help to see all valid stats.
    /// </summary>
    public static void SettingGameStats(player player, string[] args)
    {
        if (args[0] == "") return;

        var stat = args[0].ToLower();

        if (stat == "help")
        {
            GameStatsHelp(player);
            return;
        }

        if (args.Length < 2) return;

        var prop = FindProperty(Globals.GAME_STATS, stat);
        if (prop == null)
        {
            player.DisplayTimedTextTo(3.0f, $"{Colors.COLOR_YELLOW_ORANGE}No valid stat found for: |r{Colors.HighlightString(stat)} {Colors.COLOR_YELLOW_ORANGE}— try ?stat help|r");
            return;
        }

        if (!int.TryParse(args[1], out int val))
        {
            player.DisplayTimedTextTo(3.0f, $"{Colors.COLOR_YELLOW_ORANGE}Invalid value:|r {Colors.HighlightString(args[1])}");
            return;
        }

        CommandsManager.ResolvePlayerId(args.Length > 2 ? args[2] : "", kitty =>
        {
            if (kitty == null) return;
            prop.SetValue(kitty.SaveData.GameStats, val);
            player.DisplayTimedTextTo(3.0f,
                $"{Colors.COLOR_YELLOW_ORANGE}Set {Colors.HighlightString(stat)} {Colors.COLOR_YELLOW_ORANGE}to|r {Colors.HighlightString(val.ToString())} {Colors.COLOR_YELLOW_ORANGE}for|r {Colors.PlayerNameColored(kitty.Player)}");
            MultiboardUtil.RefreshMultiboards();
        });
    }

    /// <summary>
    /// Sets the specified game time for the resolved player. Args: [time] [value] [player?]. Use ?time help to see all valid times.
    /// </summary>
    public static void SettingGameTimes(player player, string[] args)
    {
        if (args[0] == "") return;

        var roundTime = args[0].ToLower();

        if (roundTime == "help")
        {
            GameTimesHelp(player);
            return;
        }

        if (args.Length < 2) return;

        var prop = FindProperty(Globals.GAME_TIMES, roundTime);
        if (prop == null)
        {
            player.DisplayTimedTextTo(3.0f, $"{Colors.COLOR_YELLOW_ORANGE}No valid time found for: |r{Colors.HighlightString(roundTime)} {Colors.COLOR_YELLOW_ORANGE}— try ?time help|r");
            return;
        }

        if (!float.TryParse(args[1], out float val))
        {
            player.DisplayTimedTextTo(3.0f, $"{Colors.COLOR_YELLOW_ORANGE}Invalid value:|r {Colors.HighlightString(args[1])}");
            return;
        }

        CommandsManager.ResolvePlayerId(args.Length > 2 ? args[2] : "", kitty =>
        {
            if (kitty == null) return;
            prop.SetValue(kitty.SaveData.RoundTimes, val);
            player.DisplayTimedTextTo(3.0f,
                $"{Colors.COLOR_YELLOW_ORANGE}Set {Colors.HighlightString(roundTime)} {Colors.COLOR_YELLOW_ORANGE}to|r {Colors.HighlightString(val.ToString())} {Colors.COLOR_YELLOW_ORANGE}for|r {Colors.PlayerNameColored(kitty.Player)}{Colors.COLOR_RESET}");
            MultiboardUtil.RefreshMultiboards();
        });
    }

    /// <summary>
    /// Gets the game stats of the passed Kitty and displays them to the player.
    /// </summary>
    public static void GetAllGameStats(player player, Kitty kitty)
    {
        if (!Globals.ALL_PLAYERS.Contains(kitty.Player)) return;
        _sb.Clear();
        foreach (var property in kitty.SaveData.GameStats.GetType().GetProperties())
        {
            var value = property.GetValue(kitty.SaveData.GameStats);
            _sb.Append($"{Colors.COLOR_YELLOW_ORANGE}{Utility.FormatAwardName(property.Name)}{Colors.COLOR_RESET}: {value}\n");
        }
        player.DisplayTimedTextTo(15.0f, $"{Colors.COLOR_YELLOW}Game stats for {Colors.PlayerNameColored(kitty.Player)}:\n{Colors.HighlightString(_sb.ToString())}{Colors.COLOR_RESET}", 0, 10);
    }

    /// <summary>
    /// Gets the personal bests of the passed Kitty and displays them to the player.
    /// </summary>
    public static void GetAllPersonalBests(player player, Kitty kitty)
    {
        if (!Globals.ALL_PLAYERS.Contains(kitty.Player)) return;
        _sb.Clear();
        var personalBests = kitty.SaveData.PersonalBests;
        foreach (var property in personalBests.GetType().GetProperties())
        {
            var value = property.GetValue(personalBests);
            _sb.Append($"{Colors.COLOR_YELLOW_ORANGE}{Utility.FormatAwardName(property.Name)}{Colors.COLOR_RESET}: {value}\n");
        }
        player.DisplayTimedTextTo(15.0f, $"{Colors.COLOR_YELLOW}Personal bests for {Colors.PlayerNameColored(kitty.Player)}:\n{Colors.HighlightString(_sb.ToString())}{Colors.COLOR_RESET}");
    }

    /// <summary>
    /// Gets the game times of the passed Kitty, optionally filtered by difficulty, and displays them to the player.
    /// </summary>
    public static void GetAllGameTimes(player player, Kitty kitty, string difficultyArg)
    {
        if (!Globals.ALL_PLAYERS.Contains(kitty.Player)) return;
        _sb.Clear();

        var properties = Globals.GAME_TIMES.GetType().GetProperties()
            .Where(p => string.IsNullOrEmpty(difficultyArg) || p.Name.ToLower().Contains(difficultyArg.ToLower()))
            .OrderBy(p => GetRoundNumber(p.Name));

        foreach (var property in properties)
        {
            var value = property.GetValue(kitty.SaveData.RoundTimes);

            string color = property.Name.Contains("Normal")     ? Colors.COLOR_YELLOW :
                           property.Name.Contains("Hard")       ? Colors.COLOR_RED :
                           property.Name.Contains("Impossible") ? Colors.COLOR_DARK_RED :
                           Colors.COLOR_YELLOW_ORANGE;

            _sb.Append($"{color}{Utility.FormatAwardName(property.Name)}{Colors.COLOR_RESET}: {Utility.ConvertFloatToTimeInt((float)value)}\n");
        }

        player.DisplayTimedTextTo(15.0f, $"{Colors.COLOR_YELLOW}Game times for {Colors.PlayerNameColored(kitty.Player)}:\n{Colors.HighlightString(_sb.ToString())}{Colors.COLOR_RESET}", 0, 0);
    }

    /// <summary>
    /// Gets the kibble currency info of the passed Kitty and displays it to the player.
    /// </summary>
    public static void GetKibbleCurrencyInfo(player player, Kitty kitty)
    {
        if (!Globals.ALL_PLAYERS.Contains(kitty.Player)) return;
        _sb.Clear();
        var kibbleCurrency = kitty.SaveData.KibbleCurrency;
        foreach (var property in kibbleCurrency.GetType().GetProperties())
        {
            var value = property.GetValue(kibbleCurrency);
            _sb.Append($"{Colors.COLOR_YELLOW_ORANGE}{Utility.FormatAwardName(property.Name)}{Colors.COLOR_RESET}: {value}\n");
        }
        var nameColored = Colors.PlayerNameColored(kitty.Player);
        player.DisplayTimedTextTo(15.0f, $"{Colors.COLOR_YELLOW}Overall Kibble Info|r ({nameColored})\n{Colors.HighlightString(_sb.ToString())}\n{Colors.COLOR_YELLOW}Current Game Info:|r ({nameColored})\n{CurrentKibbleInfo(kitty)}{Colors.COLOR_RESET}");
    }

    private static void AwardAll(player target)
    {
        if (!Globals.ALL_KITTIES.ContainsKey(target)) return;
        foreach (var category in Globals.GAME_AWARDS_SORTED.GetType().GetProperties())
        {
            var subCategory = category.GetValue(Globals.GAME_AWARDS_SORTED);
            foreach (var property in subCategory.GetType().GetProperties())
                AwardManager.GiveReward(target, property.Name);
        }
    }

    private static void AwardingHelp(player player, string filter = "")
    {
        _sb.Clear();
        int i = 0;
        int y = filter == "" ? 10 : 0;
        foreach (var category in Globals.GAME_AWARDS_SORTED.GetType().GetProperties())
        {
            var color = CategoryColors[i++ % CategoryColors.Length];
            var subCategory = category.GetValue(Globals.GAME_AWARDS_SORTED);
            _inner.Clear();
            foreach (var awd in subCategory.GetType().GetProperties())
            {
                if (string.IsNullOrEmpty(filter) || awd.Name.ToLower().Contains(filter.ToLower()))
                    _inner.Append(awd.Name).Append(' ');
            }
            if (_inner.Length > 0)
                _sb.Append($"{color}[{category.Name}]|r {_inner}\n");
        }
        player.DisplayTimedTextTo(15.0f, $"{Colors.COLOR_GOLD}Awards by Category:\n{_sb}", 0, y);
    }

    private static void GameStatsHelp(player player)
    {
        _sb.Clear();
        foreach (var property in Globals.GAME_STATS.GetType().GetProperties())
            _sb.Append(property.Name).Append(", ");
        player.DisplayTimedTextTo(15.0f, $"{Colors.COLOR_YELLOW_ORANGE}Valid game stats: {Colors.HighlightString(_sb.ToString())}");
    }

    private static void GameTimesHelp(player player)
    {
        _sb.Clear();
        foreach (var property in Globals.GAME_TIMES.GetType().GetProperties())
            _sb.Append(property.Name).Append(", ");
        player.DisplayTimedTextTo(15.0f, $"{Colors.COLOR_YELLOW_ORANGE}Valid game times: {Colors.HighlightString(_sb.ToString())}");
    }

    private static string CurrentKibbleInfo(Kitty kitty)
    {
        _inner.Clear();
        _inner.Append($"{Colors.COLOR_YELLOW_ORANGE}Collected:|r {kitty.CurrentStats.CollectedKibble}\n");
        _inner.Append($"{Colors.COLOR_YELLOW_ORANGE}Jackpots:|r {kitty.CurrentStats.CollectedJackpots}\n");
        _inner.Append($"{Colors.COLOR_YELLOW_ORANGE}Super Jackpots:|r {kitty.CurrentStats.CollectedSuperJackpots}\n");
        return _inner.ToString();
    }

    private static string FindAwardName(string input)
    {
        foreach (var category in Globals.GAME_AWARDS_SORTED.GetType().GetProperties())
        {
            var subCategory = category.GetValue(Globals.GAME_AWARDS_SORTED);
            foreach (var awd in subCategory.GetType().GetProperties())
                if (awd.Name.ToLower() == input) return awd.Name;
        }
        return null;
    }

    private static PropertyInfo FindProperty(object target, string nameLower)
    {
        foreach (var prop in target.GetType().GetProperties())
            if (prop.Name.ToLower() == nameLower) return prop;
        return null;
    }

    private static int GetRoundNumber(string propertyName)
    {
        if (propertyName.Contains("One"))   return 1;
        if (propertyName.Contains("Two"))   return 2;
        if (propertyName.Contains("Three")) return 3;
        if (propertyName.Contains("Four"))  return 4;
        if (propertyName.Contains("Five"))  return 5;
        return int.MaxValue;
    }
}

File: src/Events/Commands/AwardingCmds.ts
import { Kitty } from 'src/Game/Entities/Kitty/Kitty'
import { Globals } from 'src/Global/Globals'
import { AwardManager } from 'src/Rewards/Rewards/AwardManager'
import { GameAwardsDataSorted } from 'src/SaveSystem2.0/MAKE REWARDS HERE/SaveObjects/GameAwardsDataSorted'
import { CustomStatFrame } from 'src/UI/CustomStatFrame'
import { MultiboardUtil } from 'src/UI/Multiboard/MultiboardUtil'
import { Colors } from 'src/Utility/Colors/Colors'
import { ColorUtils } from 'src/Utility/Colors/ColorUtils'
import { isNullOrEmpty } from 'src/Utility/StringUtils'
import { Utility } from 'src/Utility/Utility'
import { MapPlayer } from 'w3ts'
import { KibbleCurrency } from '../../SaveSystem2.0/MAKE REWARDS HERE/SaveObjects/GameCurrency'
import { GameStatsData } from '../../SaveSystem2.0/MAKE REWARDS HERE/SaveObjects/GameStatsData'
import { PersonalBests } from '../../SaveSystem2.0/MAKE REWARDS HERE/SaveObjects/PersonalBests'
import { RoundTimesData } from '../../SaveSystem2.0/MAKE REWARDS HERE/SaveObjects/RoundTimesData'

export class AwardingCmds {
    /// <summary>
    /// Awards the owning player of the selected <param name="player"/> unit with the given reward input. Use ?award help to see all valid awards.
    /// </summary>
    /// <param name="player"></param>
    /// <param name="command"></param>
    public static Awarding(player: MapPlayer, args: string[]): void {
        const award = args[0].toLowerCase()
        const selectedUnit = CustomStatFrame.SelectedUnit.get(player)!
        const selectedPlayer = selectedUnit.owner

        if (args[0] === '') return

        if (award.toLowerCase() === 'help') {
            AwardingCmds.AwardingHelp(player)
            return
        }

        if (award.toLowerCase() === 'all') {
            AwardingCmds.AwardAll(player)
            return
        }

        for (const category of Object.keys(Globals.GAME_AWARDS_SORTED)) {
            const subCategory = Globals.GAME_AWARDS_SORTED[category as keyof GameAwardsDataSorted]
            for (const awd of Object.keys(subCategory)) {
                const awardString = awd.toLowerCase()
                const inputAward = award.toLowerCase()

                // Exact match
                if (awardString === inputAward) {
                    AwardManager.GiveReward(selectedPlayer, awd)
                    return
                }
            }
        }

        player.DisplayTimedTextTo(
            3.0,
            `${Colors.COLOR_YELLOW_ORANGE}No valid award found for input: |r${ColorUtils.HighlightString(award)} ${Colors.COLOR_YELLOW_ORANGE}try using ?award help|r`
        )
    }

    private static AwardingHelp = (player: MapPlayer) => {
        let combined = ''

        for (const category of Object.keys(Globals.GAME_AWARDS_SORTED)) {
            const subCategory = Globals.GAME_AWARDS_SORTED[category as keyof GameAwardsDataSorted]
            for (const awd of Object.keys(subCategory)) {
                combined += awd + ', '
            }
        }

        player.DisplayTimedTextTo(
            15.0,
            `${Colors.COLOR_YELLOW_ORANGE}Valid awards: ${ColorUtils.HighlightString(combined)}`
        )
    }

    private static AwardAll = (player: MapPlayer) => {
        for (const category of Object.keys(Globals.GAME_AWARDS_SORTED)) {
            const subCategory = Globals.GAME_AWARDS_SORTED[category as keyof GameAwardsDataSorted]
            for (const property of Object.keys(subCategory)) {
                AwardManager.GiveReward(player, property)
            }
        }
    }

    /// <summary>
    /// Sets the specified game stat for the selected player. Use ?gamestats help to see all valid game stats.
    /// </summary>
    /// <param name="player"></param>
    /// <param name="command"></param>
    public static SettingGameStats = (player: MapPlayer, args: string[]) => {
        const stats = args[0].toLowerCase()
        const selectedUnit = CustomStatFrame.SelectedUnit.get(player)!
        const selectedPlayer = selectedUnit.owner

        if (args[0] === '') return

        if (stats.toLowerCase() === 'help') {
            AwardingCmds.GameStatsHelp(player)
            return
        }

        if (args.length < 2) return

        const value = args[1]

        // Search properties for the name.. If it doesnt exist, say invalid game stat.
        // Then check if the value is actually a proper value.
        for (const prop of Object.keys(Globals.GAME_STATS)) {
            if (prop.toLowerCase() === stats.toLowerCase()) {
                let val: number
                if (!(val = S2I(value)!)) {
                    player.DisplayTimedTextTo(
                        3.0,
                        `${Colors.COLOR_YELLOW_ORANGE}Invalid value:|r ${ColorUtils.HighlightString(value.toString())}`
                    )
                    return
                }

                Globals.ALL_KITTIES.get(selectedPlayer)!.SaveData.GameStats[prop as keyof GameStatsData] = val

                player.DisplayTimedTextTo(
                    3.0,
                    `${Colors.COLOR_YELLOW_ORANGE}Set ${ColorUtils.HighlightString(stats)} ${Colors.COLOR_YELLOW_ORANGE}to|r ${ColorUtils.HighlightString(val.toString())} ${Colors.COLOR_YELLOW_ORANGE}for|r ${ColorUtils.PlayerNameColored(selectedPlayer)}`
                )
                MultiboardUtil.RefreshMultiboards()
                return
            }
        }
    }

    private static GameStatsHelp = (player: MapPlayer) => {
        let combined = ''
        for (const property of Object.keys(Globals.GAME_STATS)) {
            combined += property + ', '
        }
        player.DisplayTimedTextTo(
            15.0,
            `${Colors.COLOR_YELLOW_ORANGE}Valid game stats: ${ColorUtils.HighlightString(combined)}`
        )
    }

    public static SettingGameTimes = (player: MapPlayer, args: string[]) => {
        const roundTime = args[0].toLowerCase()
        const selectedUnit = CustomStatFrame.SelectedUnit.get(player)!
        const selectedPlayer = selectedUnit.owner

        if (args[0] === '') return

        if (roundTime.toLowerCase() === 'help') {
            AwardingCmds.GameTimesHelp(player)
            return
        }

        if (args.length < 2) return

        const value = args[1]

        // Search properties for the name.. If it doesnt exist, say invalid game stat.
        for (const prop of Object.keys(Globals.GAME_TIMES)) {
            if (prop.toLowerCase() === roundTime.toLowerCase()) {
                let val: number
                if (!(val = S2I(value)!)) {
                    player.DisplayTimedTextTo(
                        3.0,
                        `${Colors.COLOR_YELLOW_ORANGE}Invalid value:|r ${ColorUtils.HighlightString(value.toString())}`
                    )
                    return
                }

                Globals.ALL_KITTIES.get(selectedPlayer)!.SaveData.RoundTimes[prop as keyof RoundTimesData] = val

                player.DisplayTimedTextTo(
                    3.0,
                    `${Colors.COLOR_YELLOW_ORANGE}Set ${ColorUtils.HighlightString(roundTime)} ${Colors.COLOR_YELLOW_ORANGE}to|r ${ColorUtils.HighlightString(val.toString())} ${Colors.COLOR_YELLOW_ORANGE}for|r ${ColorUtils.PlayerNameColored(selectedPlayer)}${Colors.COLOR_RESET}`
                )
                MultiboardUtil.RefreshMultiboards()
                return
            }
        }
    }

    private static GameTimesHelp = (player: MapPlayer) => {
        let combined = ''
        for (const property of Object.keys(Globals.GAME_TIMES)) {
            combined += property + ', '
        }
        player.DisplayTimedTextTo(
            15.0,
            `${Colors.COLOR_YELLOW_ORANGE}Valid game times: ${ColorUtils.HighlightString(combined)}`
        )
    }

    /// <summary>
    /// Gets the game stats of the passed Kitty obj, and displays it to the player.
    /// </summary>
    /// <param name="player"></param>
    /// <param name="kitty"></param>
    public static GetAllGameStats = (player: MapPlayer, kitty: Kitty) => {
        if (!Globals.ALL_PLAYERS.includes(player)) return
        let combined = ''
        for (const property of Object.keys(Globals.GAME_STATS)) {
            const value = Globals.ALL_KITTIES.get(player)!.SaveData.GameStats[property as keyof GameStatsData]
            combined += `${Colors.COLOR_YELLOW_ORANGE}${Utility.FormatAwardName(property)}${Colors.COLOR_RESET}: ${value}\n`
        }
        player.DisplayTimedTextTo(
            15.0,
            `${Colors.COLOR_YELLOW}Game stats for ${ColorUtils.PlayerNameColored(player)}:\n${ColorUtils.HighlightString(combined)}${Colors.COLOR_RESET}`
        )
    }

    /// <summary>
    /// Gets the best personal bests of the passed Kitty obj, and displays it to the player.
    /// </summary>
    /// <param name="player"></param>
    /// <param name="kitty"></param>
    public static GetAllPersonalBests = (player: MapPlayer, kitty: Kitty) => {
        if (!Globals.ALL_PLAYERS.includes(kitty.Player)) return
        let combined = ''
        const personalBests = kitty.SaveData.PersonalBests
        for (const property of Object.keys(personalBests)) {
            const value = personalBests[property as keyof PersonalBests]
            combined += `${Colors.COLOR_YELLOW_ORANGE}${Utility.FormatAwardName(property)}${Colors.COLOR_RESET}: ${value}\n`
        }
        player.DisplayTimedTextTo(
            15.0,
            `${Colors.COLOR_YELLOW}Personal bests for ${ColorUtils.PlayerNameColored(kitty.Player)}:\n${ColorUtils.HighlightString(combined)}${Colors.COLOR_RESET}`
        )
    }

    /// <summary>
    /// Gets the best game times of the passed Kitty obj, and displays it to the player.
    /// </summary>
    /// <param name="player"></param>
    /// <param name="kitty"></param>
    public static GetAllGameTimes = (player: MapPlayer, kitty: Kitty, difficultyArg: string) => {
        if (!Globals.ALL_PLAYERS.includes(kitty.Player)) return
        let combined = ''

        // Previously this wasn't sorted by round number, so i had to hard code the order with one two three etc.. but ye its sorted now
        const properties = Object.keys(Globals.GAME_TIMES)
            .filter(p => isNullOrEmpty(difficultyArg) || p.toLowerCase().includes(difficultyArg.toLowerCase()))
            .sort((a, b) => AwardingCmds.GetRoundNumber(a) - AwardingCmds.GetRoundNumber(b))

        for (const property of properties) {
            const value = kitty.SaveData.RoundTimes[property as keyof RoundTimesData] || 0

            const color = property.includes('Normal')
                ? Colors.COLOR_YELLOW
                : property.includes('Hard')
                  ? Colors.COLOR_RED
                  : property.includes('Impossible')
                    ? Colors.COLOR_DARK_RED
                    : Colors.COLOR_YELLOW_ORANGE // Default fallback

            combined += `${color}${Utility.FormatAwardName(property)}${Colors.COLOR_RESET}: ${Utility.ConvertFloatToTimeInt(value)}\n`
        }

        player.DisplayTimedTextTo(
            15.0,
            `${Colors.COLOR_YELLOW}times: for: Game ${ColorUtils.PlayerNameColored(kitty.Player)}:\n${ColorUtils.HighlightString(combined)}${Colors.COLOR_RESET}`
        )
    }

    private static GetRoundNumber = (propertyName: string) => {
        if (propertyName.includes('One')) return 1
        if (propertyName.includes('Two')) return 2
        if (propertyName.includes('Three')) return 3
        if (propertyName.includes('Four')) return 4
        if (propertyName.includes('Five')) return 5

        return math.maxinteger
    }

    /// <summary>
    /// Gets the kibble currency of the passed Kitty obj, and displays it to the player.
    /// </summary>
    /// <param name="player"></param>
    /// <param name="kitty"></param>
    public static GetKibbleCurrencyInfo = (player: MapPlayer, kitty: Kitty) => {
        if (!Globals.ALL_PLAYERS.includes(kitty.Player)) return
        let combined = ''
        const kibbleCurrency = kitty.SaveData.KibbleCurrency
        for (const key of Object.keys(kibbleCurrency)) {
            const value = kibbleCurrency[key as keyof KibbleCurrency]
            combined += `${Colors.COLOR_YELLOW_ORANGE}${Utility.FormatAwardName(key)}${Colors.COLOR_RESET}: ${value}\n`
        }
        const nameColored = ColorUtils.PlayerNameColored(kitty.Player)
        player.DisplayTimedTextTo(
            15.0,
            `${Colors.COLOR_YELLOW}Kibble: Info: Overall|r (${nameColored})\n${ColorUtils.HighlightString(combined)}\n${Colors.COLOR_YELLOW}Game: Info: Current:|r (${nameColored})\n${AwardingCmds.CurrentKibbleInfo(kitty)}${Colors.COLOR_RESET}`
        )
    }

    private static CurrentKibbleInfo = (kitty: Kitty) => {
        let combined = ''
        combined += `${Colors.COLOR_YELLOW_ORANGE}Collected:|r ${kitty.CurrentStats.CollectedKibble}\n`
        combined += `${Colors.COLOR_YELLOW_ORANGE}Jackpots:|r ${kitty.CurrentStats.CollectedJackpots}\n`
        combined += `${Colors.COLOR_YELLOW_ORANGE}Super Jackpots:|r ${kitty.CurrentStats.CollectedSuperJackpots}\n`
        return combined
    }
}
